Refactor Actors performance counter emission

ilijagrbic opened the pull request:
Refactor the way we emit diagnostic data in Actors. Implement more testable abstraction in order to prepare a path for Metrics instrumentation.
- Delete old `DiagnosticManager`/`DiagnosticEventManager` and clean up old code
- Introduce `IDiagnostics`
- Implement composite pattern in `AggregatedDiagnoticEvents`
- Implement `EventSourceDiagnoticEvents` and `PerformanceCounterDiagnoticEvents`
- Replace old `DiagnoticManager` usage with new `AggregatedDiagnoticEvents`

ilijagrbic committed:
Implement TelemetryDiagnosticEvents

ilijagrbic committed:
Add ServiceContext as parameter for FabricTransportMessageHandler constructor

ilijagrbic committed:
Integrate TelemetryDiagnosticEvents into MessageHandler

ilijagrbic committed:
Change commend

ilijagrbic committed:
Improve tests

ilijagrbic committed:
Merge branch 'develop' into user/ilijagrbi/implementTelemetryDiagnosticEvents

ilijagrbic committed:
Extract telemetry mock logic into MockedTelemetryTest for reusability

ilijagrbic committed:
Merge branch 'user/ilijagrbi/implementTelemetryDiagnosticEvents' of https://github.com/microsoft/service-fabric-services-and-actors-dotnet into user/ilijagrbi/implementTelemetryDiagnosticEvents

ilijagrbic committed:
Self-review

ilijagrbic committed:
Update TelemetryDiagnosticEventsTest.cs

ilijagrbic committed:
Refactor all methods to be protected in EventSourceProvider if possible

ilijagrbic committed:
Refactor all methods to be protected in PerformanceCounterProvider if possible

ilijagrbic committed:
Remove unused methods in DiagnosticsEventManager

ilijagrbic committed:
Switch properties to protected

ilijagrbic committed:
Merge branch 'develop' into user/ilijagrbi/refactorActorsPerfCounters

ilijagrbic committed:
Fix merge

ilijagrbic committed:
Implement AgregateDiagnosticEvents

ilijagrbic committed:
Initialize PerformanceCounterDiagnosticEvents and tests

ilijagrbic committed:
Implement rest of PerformanceCounterDiagnosticEvents

ilijagrbic committed:
Fix ActorStateLoad test

ilijagrbic committed:
Merge branch 'develop' into user/ilijagrbi/refactorActorsPerfCounters

ilijagrbic committed:
Move Actor perf counters to seperate namespace

ilijagrbic committed:
Move diagnotic data and old (obsolete) diagnostic classes to seperate namespaces

ilijagrbic committed:
Revert move old (obsolete) diagnostic classes to seperate namespaces

ilijagrbic committed:
Move old (obsolete) diagnostic classes to seperate namespace

ilijagrbic committed:
Update AgregateDiagnosticEvents.cs

ilijagrbic committed:
Extract TestActor and optimize verify method calls in PerfCounter tests

ilijagrbic committed:
Extract ActorMethodInfo

ilijagrbic committed:
Enable mocking of ActorFrameworkEventSource

ilijagrbic committed:
Init EventSourceDiagnosticEvents and tests

ilijagrbic committed:
Merge branch 'develop' into user/ilijagrbi/refactorActorsPerfCounters

ilijagrbic committed:
Implement EventSource events except for ActorMethod ones

ilijagrbic committed:
Implement ActorMethod tracing

ilijagrbic committed:
Init actorMethodInfo

ilijagrbic committed:
refactor InitializeActorMethodInfo

ilijagrbic committed:
Fix build

ilijagrbic committed:
Fix test

ilijagrbic committed:
Rename test

ilijagrbic committed:
Fix linux test

ilijagrbic committed:
Implement pendingMethodCall logic

ilijagrbic committed:
Use ItExpr insteade of It

ilijagrbic committed:
Move ActorManagerTests

ilijagrbic committed:
Initialize DiagnosticEvents in ActorManager

ilijagrbic committed:
Use IDiagnosticEvents in ActorManager

ilijagrbic committed:
Use IDiagnosticEvents in ActorStateManager

ilijagrbic committed:
Use IDiagnosticEvents in ActorService

ilijagrbic committed:
Use IDiagnosticEvents in ActorBase

ilijagrbic committed:
Use IDiagnosticEvents in RemotingDispatcher

ilijagrbic committed:
Delete old IDiagnosticsManager

ilijagrbic committed:
Fix ActorManager constructor tests

ilijagrbic committed:
Delete unused methods in PerfCounterProvider

ilijagrbic committed:
Update ActorManagerTests.cs

ilijagrbic committed:
Update ActorManagerTests.cs

ilijagrbic committed:
Delete unused fields/structs in DiagnosticsManagerActorContext

ilijagrbic committed:
Restore ActorManager static field

ilijagrbic committed:
Disable Linux code path for ServiceFabricEventSource in EventSource tests

ilijagrbic commented on src/Microsoft.ServiceFabric.Actors/Diagnostics/AggregatedDiagnosticEvents.cs:
```diff
@@ -0,0 +1,176 @@
+        }
+        public void AcquireActorLockFailed(DiagnosticsManagerActorContext diagnosticContext)
+        {
+            Interlocked.Decrement(ref diagnosticContext.PendingActorMethodCalls);
```
This is how PendingMethodCalls were counted in the previous DiagnosticsManager. Think about if there is a better way to do this.

olegsych replied:
This definitely smells like an unrelated responsibility in the context of the AggregatedDiagnosticEvents. 
This could be a distinct class implementating of the IDiagnosticEvents with the single responsibility of tracking the PendingActorMethodCalls, but since only two methods would be doing the actual work, it feels like an overkill. Why couldn't this be done by the caller?

ilijagrbic replied:
Thanks for the comment, Oleg, this is a good point. Another reason I didn't want to implement it as an another `IDiagnoticEvents` since in that case logic for other `IDiagnoticEvents` would depend on their ordering in the `AggregatedDiagnoticEvents` list, which would me an implicit constraint which could easily be broken by accident. I think we can move this to be done by the caller, it would be just a bit harder to test, but I will look into moving it there.

ilijagrbic replied:
I'm planning to move the logic for this into the `DiagnosticsManagerActorContext` class, and unit test it that way. I will probably finish that up on Monday. :)

ilijagrbic replied:
In the end I moved the logic to the `DiagnosticsManagerActorContext`, which also allowed to me to remove some not needed APIs from the `IDiagnoticEvents` (`AcquireActorLockStart` and `AcquireActorLockFailed`) which were only needed in order to increment/decrement pending calls.

ilijagrbic commented on src/Microsoft.ServiceFabric.Actors/Diagnostics/Util.cs:
```diff
@@ -5,9 +5,13 @@
-    internal struct ActivationDiagnosticData
+    internal class Util
     {
-        internal bool IsActivationEvent;
-        internal ActorId ActorId;
+        public static long GetInterfaceMethodKey(uint interfaceId, uint methodId)
+        {
+            var key = (ulong)methodId;
+            key = key | (ulong)interfaceId << 32;
+            return (long)key;
+        }
```
Consider a better place for this logic.

ilijagrbic commented on src/Microsoft.ServiceFabric.Actors/Runtime/ActorManager.cs:
```diff
@@ -34,26 +35,40 @@ internal sealed class ActorManager : IActorManager
 
+        private readonly IDiagnosticEvents diagnosticEvents;
+        private readonly IClock clock;
+        private static Func<ActorService, IClock, IDiagnosticEvents> createDiagnosticEvents = (actorService, clock) =>
```
We need this static field in order to mock `diagnosticEvents` in certain test paths, where `ActorManager` is not injected but rather created during the process.

ilijagrbic commented on test/Microsoft.ServiceFabric.Actors.Tests/Diagnostics/PerformanceCounterDiagnosticEventsTest.cs:
```diff
@@ -0,0 +1,367 @@
+                {
+                    sut.SaveActorStateFinish(actorId, startTime);
+
+                    Mock.Get(actorSaveStateTimeCounterWriter).Verify(p => p.UpdateCounterValue(It.Is<ActorStateDiagnosticData>(data => data.OperationTime.Value.TotalMilliseconds == operationDurationMillis && data.ActorId == actorId)), Times.Once);
```
TODO: Use Equals here instead of manually checking fields.

ilijagrbic committed:
address self-review

ilijagrbic committed:
Rename sut in ActorManagerTest

ilijagrbic committed:
Fix ActorStateManagerTest

ilijagrbic committed:
Address copilot comments

olegsych commented on src/Microsoft.ServiceFabric.Actors/Diagnostics/Data/ActorMethodDiagnosticData.cs:
We should keep the folder structure consistent with the namespace structure for maintainability.
With only three types, I don't think we have enough to justify the cognitive overhead of an extra sub-namespace/folder for Data. .NET used to have an FxCop rule advising against namespaces with fewer than 5 types, if memory serves me well. To be on the safe side, my rule of thumb to decide that a new namespace is needed once the number of candidate types reaches 7.

ilijagrbic replied:
Good point, I will move them to the parent namespace, thanks.

olegsych commented on src/Microsoft.ServiceFabric.Actors/Diagnostics/PerformanceCounters/ActorMethodExceptionFrequencyCounterWriter.cs:
```diff
@@ -11,14 +11,15 @@ namespace Microsoft.ServiceFabric.Actors.Diagnostics
     // frequency at which a particular actor method throws exceptions.
     internal class ActorMethodExceptionFrequencyCounterWriter : FabricBaselessPerformanceCounterWriter
     {
+        protected ActorMethodExceptionFrequencyCounterWriter() { }
```
Could you double-check the code style? This constructor is missing a blank line.

olegsych commented on src/Microsoft.ServiceFabric.Actors/Diagnostics/PerformanceCounters/ActorLockContentionCounterWriter.cs:
```diff
@@ -11,14 +11,16 @@ namespace Microsoft.ServiceFabric.Actors.Diagnostics
     // number of pending actor calls that are waiting for the actor lock.
```
We have enough types to justify a separate folder for PerfCounters, but let's add a matching sub-namespace

olegsych commented on src/Microsoft.ServiceFabric.Actors/Diagnostics/PerformanceCounters/ActorLockContentionCounterWriter.cs:
```diff
@@ -11,14 +11,16 @@ namespace Microsoft.ServiceFabric.Actors.Diagnostics
     // number of pending actor calls that are waiting for the actor lock.
     internal class ActorLockContentionCounterWriter : FabricBaselessPerformanceCounterWriter
     {
+        protected ActorLockContentionCounterWriter() { }
```
Why do we need protected constructors? I'm guessing we need to dynamically subclass them in Moq, but doesn't the InternalsVisibleTo already cover it?

ilijagrbic replied:
I think that `InternalsVisibleTo` would allow `Moq` to access internal methods/fields. However, since we defined a constructor with parameters, no default constructor without parameters is present in this class and I needed it since I didn't want to pass `FabricPerformanceCounterSetInstance` in order to create a Mock object. Anyways that's why I defined this protected constructor.

olegsych commented on src/Microsoft.ServiceFabric.Actors/Diagnostics/PerformanceCounters/ActorLockContentionCounterWriter.cs:
```diff
@@ -11,14 +11,16 @@ namespace Microsoft.ServiceFabric.Actors.Diagnostics
         }
 
-        internal void UpdateCounterValue(PendingActorMethodDiagnosticData pendingMethodData)
+        internal virtual void UpdateCounterValue(PendingActorMethodDiagnosticData pendingMethodData)
```
I wonder if we should add a one-line comment to the top of the class to explain that it's not sealed to enable dynamic mocking.

ilijagrbic replied:
Good idea, I will add a comment explaining this.

olegsych commented on src/Microsoft.ServiceFabric.Actors/Diagnostics/ActorMethodInfo.cs:
```diff
@@ -0,0 +1,24 @@
+
+namespace Microsoft.ServiceFabric.Actors.Diagnostics
+{
+    internal class ActorMethodInfo
```
Could you omit `internal` in new code? C# is a very well-designed language and defaults,, such as types are internal unless specified otherwise, help to reduce the noise and make the code more readable. 

On the other hand, we should make classes sealed as much as possible to communicate the whether we intend them to be subclassed.

olegsych commented on src/Microsoft.ServiceFabric.Actors/Diagnostics/ActorMethodInfo.cs:
```diff
@@ -0,0 +1,24 @@
+{
+    internal class ActorMethodInfo
+    {
+        internal ActorMethodInfo() { }
```
This constructor appears to be redundant because C# should provide an empty constructor by default.

ilijagrbic replied:
Maybe I'm wrong about this but I think default constructor is provided only if there are no other constructors defined. But I will look into if I can remove the usage of this altogether.

ilijagrbic replied:
Since this class is really simple, I have moved the logic from this constructor to the caller and now the class just sets fields to passed parameters. I also removed constructor without parameters and set properties.

olegsych commented on src/Microsoft.ServiceFabric.Actors/Diagnostics/ActorMethodInfo.cs:
```diff
@@ -0,0 +1,24 @@
+            MethodSignature = methodInfo.ToString();
+        }
+
+        internal string MethodName { get; set; }
```
Can we define these as read-only fields instead of read-write properties? The design guideline recommending properties shouldn't apply to internal APIs.

olegsych commented on src/Microsoft.ServiceFabric.Actors/Diagnostics/AggregatedDiagnosticEvents.cs:
```diff
@@ -0,0 +1,176 @@
+{
+    internal class AggregatedDiagnosticEvents : IDiagnosticEvents
+    {
+        readonly IEnumerable<IDiagnosticEvents> diagnosticEvents;
```
Could you double-check for blank lines between type members?

olegsych commented on src/Microsoft.ServiceFabric.Actors/Diagnostics/AggregatedDiagnosticEvents.cs:
```diff
@@ -0,0 +1,176 @@
+        {
+            _ = diagnosticEvents ?? throw new ArgumentNullException(nameof(diagnosticEvents));
+            if (diagnosticEvents.Any(d => d == null))
+            {
```
I would love to eliminate curly braces around single line statements. The bloat is particularly noticeable in a class like this.

olegsych commented on src/Microsoft.ServiceFabric.Actors/Diagnostics/EventSourceDiagnosticEvents.cs:
```diff
@@ -0,0 +1,189 @@
+        public EventSourceDiagnosticEvents(ActorFrameworkEventSource eventSource, IClock clock, ServiceContext serviceContext, ActorMethodFriendlyNameBuilder nameBuilder, ActorTypeInformation typeInfo)
+        {
+            this.eventSource = eventSource ?? throw new ArgumentNullException(nameof(eventSource));
+            this.serviceContext = serviceContext;
```
Shouldn't we throw on null ServiceContext as well?

ilijagrbic replied:
Yeah, that is correct, I will add a check for this.

olegsych commented on src/Microsoft.ServiceFabric.Actors/Diagnostics/EventSourceDiagnosticEvents.cs:
```diff
@@ -0,0 +1,189 @@
+            this.actorMethodInfo = InitializeActorMethodInfo(nameBuilder, typeInfo);
+        }
+
+        private Dictionary<long, ActorMethodInfo> InitializeActorMethodInfo(ActorMethodFriendlyNameBuilder nameBuilder, ActorTypeInformation typeInfo)
```
Same ask to avoid specifying private for type members, which are private by default.

olegsych commented on src/Microsoft.ServiceFabric.Actors/Diagnostics/EventSourceDiagnosticEvents.cs:
```diff
@@ -0,0 +1,189 @@
+        {
+            var actorMethodInfo = new Dictionary<long, ActorMethodInfo>();
+
+            foreach (var actorInterfaceType in typeInfo.InterfaceTypes)
```
What is the actual type of 'var` here?

ilijagrbic replied:
It's `System.Type`, I thought it is self-explanatory from the variable name, but I will add explicit types here to make it more clear.

olegsych commented on src/Microsoft.ServiceFabric.Actors/Diagnostics/EventSourceDiagnosticEvents.cs:
```diff
@@ -0,0 +1,189 @@
+            this.actorMethodInfo = InitializeActorMethodInfo(nameBuilder, typeInfo);
+        }
+
+        private Dictionary<long, ActorMethodInfo> InitializeActorMethodInfo(ActorMethodFriendlyNameBuilder nameBuilder, ActorTypeInformation typeInfo)
```
Does InitializeActorMethodInfo initialize this.actorMethodInfo or only builds it? Can this method be static?

ilijagrbic replied:
Good catch, initially it was initializing the field, but I refactored it in order to be able to make the field `readonly`. Anyways, I will rename the method and change it to be static. :)

olegsych commented on src/Microsoft.ServiceFabric.Actors/Diagnostics/EventSourceDiagnosticEvents.cs:
```diff
@@ -0,0 +1,189 @@
+
+        public void AcquireActorLockFinish(PendingActorMethodDiagnosticData diagnosticData, DateTime startTime)
+        {
+            if (this.eventSource.IsPendingMethodCallsEventEnabled())
```
We can omit this. without losing any readability in small methods like this. I always omit this. and when the method starts to feel like it needs it, that's a sign to break it up.

ilijagrbic replied:
Good point, I will remove it where it is not needed (which should be everywhere except for constructor).

olegsych commented on src/Microsoft.ServiceFabric.Actors/Diagnostics/EventSourceDiagnosticEvents.cs:
```diff
@@ -0,0 +1,189 @@
+            if (e != null)
+            {
+                this.eventSource.ActorMethodThrewException(
+                   e.ToString(),
```
Ugh... I really dislike the convention of placing method call parameters on separate lines. As soon as we start doing this, we throw away an important complexity indicator and ignore the opportunity to improve the method design. Offhand, this method smells of "primitives obsession" and asking to take an Exception instead of a string, a DateTime instead of a long, an ActorMethodInfo instead of two strings.

ilijagrbic replied:
I agree on the number of parameters and on this method. I will refactor this method to accept `ActorMethodDiagnosticData` instead of multiple primitive parameters. However, I will keep the call to `ActorFrameworkEventSource` as it is, since that would be yet another change in an already big PR.

olegsych commented on src/Microsoft.ServiceFabric.Actors/Remoting/V2/Runtime/ActorServiceRemotingDispatcher.cs:
```diff
@@ -26,8 +26,8 @@ namespace Microsoft.ServiceFabric.Actors.Remoting.V2.Runtime
         private static readonly string TraceType = typeof(ActorServiceRemotingDispatcher).Name;
-        private readonly ActorService actorService;
-        private readonly ServiceRemotingCancellationHelper cancellationHelper;
+        private ActorService actorService;
```
Why are we making these fields writeable? If this is for tests, then readonly instance fields can be modified by reflection, only readonly static fields cannot be.

ilijagrbic replied:
Ahhh,, sorry, I was removing the `private` modifier on a bunch of fields, and I removed `readonly` by mistake here...

olegsych commented on src/Microsoft.ServiceFabric.Actors/Runtime/ActorStateManager.cs:
```diff
@@ -12,21 +12,27 @@ namespace Microsoft.ServiceFabric.Actors.Runtime
             this.actor = actor;
             this.stateProvider = actorStateProvider;
             this.stateChangeTracker = new Dictionary<string, StateMetadata>();
+            this.clock = actor.ActorService.Clock;
```
This is breaks the law of Demeter and creates a tight coupling between components. We should strive to maintain direct dependencies only. For internal classes, this would be as straightforward as declaring additional constructor parameters.

ilijagrbic replied:
The way I was thinking about it was that either I will access the `IClock/IDiagnoticEvents` here or I will inject them from `ActorService`. However, in `ActorService` I would basically have the same logic (`actor.ActorService.Clock` and `actor.Manager.DiagnosticsEvents`) or I would have to change the public API. Tight coupling already existed here, since we were accessing old `DiagnoticManager` in a similar manner.

Anyways, I see a benefit for injecting these anyways, particularly in the test code, where I had to set some fields in `ActorService` in order to test the `ActorStateManager`. So, I will inject these as constructor parameters and move the logic to access them to `ActorService`, let me know what you think about this. :)

olegsych commented on test/Microsoft.ServiceFabric.Actors.Tests/Diagnostics/PerformanceCounterDiagnosticEventsTest.cs:
```diff
@@ -0,0 +1,369 @@
+            public void ThrowsOnNullProvider()
+            {
+                var exception = Assert.Throws<ArgumentNullException>(() =>
+                {
```
Could you simplify the lambdas? Even if you don't want to place it on a single line, there is no need or benefit to the curly braces.

olegsych reviewed (commented):
Hey @ilijagrbic, this is a large change and I haven't been able to finish thoroughly reviewing it. The comments below are mostly small style and design nits. 

At a higher level, I'm not sure replacing DiagnosticsEventManager with the IDiagnosticEvents hierarchy is worth the risk of such a significant change. Other than consistency with the design we're using in other projects, what are the comparative advantages here? The old design is based on .NET delegates which already implement the composite pattern and don't require implementing AggregatedDiagnosticEvents. Otherwise, but both designs appear to produce the same number of classes and could be fully tested. Doesn't that make the delegate-based design simpler?

ilijagrbic commented:
Hey @olegsych , thank you for the review. There are couple of points I feel like are better in this design:
- There is smaller number of classes for this design. We now don't have some specific data classes only used to repack method parameters. We also don't have two `EventSourceProviders`. Perhaps total number of files seems the same since I extracted some classes into separate files, such as Util.
- Previously logic in `DiagnosticEventManager` used to change struct values in `DiagnoticContext` before passing it allong to `EventSource`/`PerfCounterProviders`.
- We removed entire `Stopwatch `logic. It was in my opinion overly complex, especially for nested method call and really not needed if replaced by the pattern we used in Remoting.
- Both patterns do essentially achieve the same thing, I just think new pattern is easier to test and mock.
- I also think there is value in consistency across our projects, especially for such a similar process. Future readers would understand it easier and would not need to wonder why we have such different architecture for the same thing.

On the other hand, I see where your concerns come from, I had similar concerns and entire change grew a bit too much I feel. We can discuss if this refactor is worth the risk and if not, we can always move on instrumenting Actors code by using the old pattern.

ilijagrbic committed:
Merge branch 'develop' into user/ilijagrbi/refactorActorsPerfCounters

ilijagrbic committed:
Fix PR comments - address namespace issues

ilijagrbic committed:
Add comment explaining why ActorFrameworkEventSouce is not sealed

ilijagrbic committed:
Seal new classes not intended for inheritance

ilijagrbic committed:
Fix build errors

ilijagrbic committed:
Update AggregatedDiagnosticEvents.cs

ilijagrbic committed:
Add throw on null serviceContext fort eventSourceEvents

ilijagrbic committed:
Shorten test exception lambdas

ilijagrbic committed:
Reintroduce mistakenly removed readonly

ilijagrbic committed:
Remove this. from DiagnoticEvents classes where not needed

ilijagrbic committed:
Update EventSourceDiagnosticEventsTest.cs

ilijagrbic committed:
Update PerformanceCounterDiagnosticEventsTest.cs

ilijagrbic committed:
Refactor ActorMethodInfo

ilijagrbic committed:
Inject IClock and IDiagnoticManager into ActorStateManager

ilijagrbic committed:
Merge branch 'develop' into user/ilijagrbi/refactorActorsPerfCounters

ilijagrbic committed:
Remove old Increment/Decrement logic for actor methods

ilijagrbic committed:
Move Incrmeent/Decrement responsibility away from IDiagnosticEvents

ilijagrbic committed:
Update DiagnosticsManagerActorContextTest.cs

ilijagrbic committed:
Refactor ActorMethodFinish signiture

ilijagrbic committed:
Refactor ActorMethodStart signiture

ilijagrbic committed:
Fix build

ilijagrbic committed:
Remove TestActor class and use MockActor

ilijagrbic committed:
Dispose of PerformanceCountersProvider

ilijagrbic committed:
Rename DiagnosticsManagerActorContext -> DiagnosticActorContext

ilijagrbic commented on test/Microsoft.ServiceFabric.Actors.Tests/Runtime/ActorManagerTests.cs:
```diff
@@ -0,0 +1,343 @@
+            actorService.InitializeInternal(friendlyNameBuilder);
+        }
+
+        public class Remainder : ActorManagerTests
```
Note: `Remainder` class contains only old tests for `ActorManager`. They are shown as new code since the file was moved to this package.

ilijagrbic committed:
Self-review

ilijagrbic committed:
Update ActorManagerTests.cs

ilijagrbic committed:
Rename MockActor

ilijagrbic committed:
Fix tests

olegsych commented on src/Microsoft.ServiceFabric.Actors/Diagnostics/PerformanceCounters/ActorLockContentionCounterWriter.cs:
```diff
@@ -3,22 +3,24 @@
         }
 
-        internal void UpdateCounterValue(PendingActorMethodDiagnosticData pendingMethodData)
+        internal virtual void UpdateCounterValue(PendingActorMethodDiagnosticData pendingMethodData)
         {
```
Forgot to mention last time. When I was thinking about how to unit test this code, I though we could use Inspector to get to the FabricPerformanceCounter obtained by the base class and use its GetValue method. This is not as invasive as making the UpdateCounterValue methods virtual and would be closer to the state-based testing, which is generally more maintainable than behavior-based testing we do with mocks. It's a pretty wide gray area though, so just a suggestion to consider.

ilijagrbic replied:
Thanks for the comment, I also think that state-based testing is a more maintainable approach. However, it would require us to introduce some changes into the PerfCounter classes. While these changes would be minor (for example initializing `FabricPerformanceCounter` in constructors without fields), I didn't feel like doing this and going through tedious release process for WindowsFabric NuGet packages is worth the improvements we would get that way. 
Also, I try to avoid interacting too much with the internals of classes that are not being tested since I think that introduces unnecessary complexity. I'll think about this some more, and we can refactor these tests if it turns out that current approach significantly impacts the maintainability of these test.

olegsych commented on src/Microsoft.ServiceFabric.Actors/Diagnostics/ActorFrameworkEventSource.cs:
```diff
@@ -11,17 +11,20 @@
 
         [NonEvent]
-        internal void ReplicaChangeRoleToPrimary(ServiceContext serviceContext)
+        internal virtual void ReplicaChangeRoleToPrimary(ServiceContext serviceContext)
```
I don't remember if I suggested before and you probably already know that we can observe event method calls using EventListener without making the "NonEvent" methods virtual. However, verifying the high-level event methods with mocks is much more concise so overall, I think this approach is *simpler*.

ilijagrbic replied:
Yeah, I have decided to use the simpler approach here. This way a boundary of what is being testes is clearer and I don't have to know anything about the internals of the `ActorFrameworkEventSource`, but a good suggestion nonetheless. :)

olegsych commented on src/Microsoft.ServiceFabric.Actors/Diagnostics/ActorMethodDiagnosticData.cs:
```diff
@@ -12,8 +12,16 @@ internal struct ActorMethodDiagnosticData
         internal RemotingListenerVersion RemotingListener;
+        internal TimeSpan? MethodExecutionTime;
+
+        internal ActorMethodDiagnosticData(ActorId actorId, long interfaceMethodKey, Exception exception, RemotingListenerVersion remotingListener)
```
Now that we have a constructor, can we make the fields read-only as well?

ilijagrbic replied:
I removed the constructor, and I will tackle this issue in the follow up PR.

olegsych commented on src/Microsoft.ServiceFabric.Actors/Diagnostics/ActorMethodDiagnosticData.cs:
```diff
@@ -12,8 +12,16 @@ internal struct ActorMethodDiagnosticData
+        internal TimeSpan? MethodExecutionTime;
+
+        internal ActorMethodDiagnosticData(ActorId actorId, long interfaceMethodKey, Exception exception, RemotingListenerVersion remotingListener)
+        {
```
It's odd that we don't have the method execution time with the exception. I could see them being mutually exclusive, but there's value in measuring the execution time even when an exception is thrown.

ilijagrbic replied:
I wanted to keep the responsibility of calculating the execution time in the respective `IDiagnotic` classes, so that's why this constructor doesn't accept that info. Ideally, I would like to change the signature of PerfCounter methods to not accept this structure, since they actually don't use most of the data from it. This would also enable me to remove the execution time from this struct and make all other fields `readonly`.

Anyways, I will remove the constructor, and I will tackle this in the follow up PR.

olegsych commented on src/Microsoft.ServiceFabric.Actors/Diagnostics/ActorMethodDiagnosticData.cs:
```diff
@@ -12,8 +12,16 @@ internal struct ActorMethodDiagnosticData
+
+        internal ActorMethodDiagnosticData(ActorId actorId, long interfaceMethodKey, Exception exception, RemotingListenerVersion remotingListener)
+        {
+            ActorId = actorId;
```
I'm not sure why we're adding this constructor, but since we're doing it, don't we want to validate the parameters? I assume we'd want to check ActorId and Exception parameters for null values.

ilijagrbic replied:
Main reason why I added this constructor is to improve readability. Using for example `new ActorMethodDiagnosticData() { actorId = ActorId, ... }` often lead to longer line length than I would like to have. I will probably revert this change as I think it raised more questions than it helped solve.

olegsych commented on src/Microsoft.ServiceFabric.Actors/Diagnostics/ActorMethodInfo.cs:
```diff
@@ -0,0 +1,23 @@
+            this.methodSignature = methodSignature;
+        }
+
+        internal string MethodName => methodName;
```
Since this class is internal, I would make MethodName and MethodSignature internal readonly fields rather than properties. In API design, properties allow changing public APIs in the future, so I prefer the simplicity of readonly fields. But you could make a case for internal APIs due to our excessive use of InternalsVisibleTo. So, if you strongly prefer properties, make them get-only and remove the fields.

ilijagrbic replied:
I was unsure about this as well and I mostly added the properties to avoid having externally visible fields, since that seemed wrong to me. Now that I look at it again, the properties seem rather artificial in this place, so I will take your suggestion and make the fields internal.

olegsych commented on src/Microsoft.ServiceFabric.Actors/Diagnostics/AggregatedDiagnosticEvents.cs:
```diff
@@ -0,0 +1,145 @@
+
+        internal AggregatedDiagnosticEvents(IEnumerable<IDiagnosticEvents> diagnosticEvents)
+        {
+            _ = diagnosticEvents ?? throw new ArgumentNullException(nameof(diagnosticEvents));
```
My OCD compels me to dislike using a throw-away variable assignment to eliminate an extra line of an if statement. I'm not sure it's rational though. Do you know if this is optimized away by the compiler?

ilijagrbic replied:
Main reason was to avoid having an extra line of an `if` statement. I'm kind of struggling to decide which of those two options were better and I got PR comments lately that `if` can be avoided this way. Anyways, I started to like the uniformity of how parameters are checked now.

olegsych commented on src/Microsoft.ServiceFabric.Actors/Diagnostics/ActorFrameworkEventSource.cs:
```diff
@@ -140,13 +143,13 @@ internal void ActorMethodStart(
         [NonEvent]
-        internal void ActorMethodStop(
+        internal virtual void ActorMethodStop(
            long methodExecutionTimeTicks,
```
I think TimeSpan to ticks conversion would be *simpler* to encapsulate here rather than in the EvenSourceDiagnosticEvents

ilijagrbic replied:
I agree with this, I will address this in a follow up PR together with other logic that should be moved into the EventSource class.

olegsych commented on test/Microsoft.ServiceFabric.Actors.Tests/Diagnostics/DiagnosticsManagerActorContextTest.cs:
```diff
@@ -0,0 +1,49 @@
+
+namespace Microsoft.ServiceFabric.Actors.Diagnostics
+{
+    public class DiagnosticsManagerActorContextTest
```
Let's make the test class match the name of the SUT.

olegsych commented on src/Microsoft.ServiceFabric.Actors/Diagnostics/DiagnosticActorContext.cs:
```diff
@@ -0,0 +1,32 @@
+
+namespace Microsoft.ServiceFabric.Actors.Diagnostics
+{
+    internal class DiagnosticActorContext
```
This reads "Context of a Diagnostic Actor". I think "ActorDiagnosticsContext" would be more accurate and consistent with the ActorBase.DiagnosticsContext property.

ilijagrbic replied:
I like this suggestion, this name was kind of a left over from the old name for this class. I will rename the class into `DiagnosticContext` though, since I think it is implied that this context is related to Actor when it is its member.

olegsych commented on src/Microsoft.ServiceFabric.Actors/Diagnostics/EventSourceDiagnosticEvents.cs:
```diff
@@ -0,0 +1,175 @@
+                return;
+            }
+
+            if (eventSource.IsActorMethodStopEventEnabled())
```
I wonder if this logic would be simpler to encapsulate inside the eventSource.ActorMethodStop.

ilijagrbic replied:
I agree, I will keep it outside of this PR, but I will address this in the future.

olegsych commented on src/Microsoft.ServiceFabric.Actors/Runtime/ActorBase.cs:
```diff
@@ -99,7 +102,7 @@ public ActorService ActorService
         internal bool IsInitialized { get; set; }
 
-        internal DiagnosticsManagerActorContext DiagnosticsContext
+        internal DiagnosticActorContext DiagnosticsContext
```
Context is a filler word. Since we're changing the property type, I would consider dropping it and naming the property simply Diagnostics.

olegsych commented on src/Microsoft.ServiceFabric.Actors/Runtime/ActorBase.cs:
```diff
@@ -120,9 +123,12 @@ internal IActorManager Manager
         internal async Task OnActivateInternalAsync()
         {
-            this.Manager.DiagnosticsEventManager.ActorOnActivateAsyncStart(this);
+            var startTime = clock.UtcNow;
```
Let's use explicit types instead of var.

olegsych commented on src/Microsoft.ServiceFabric.Actors/Runtime/ActorStateManager.cs:
```diff
@@ -12,21 +12,27 @@ namespace Microsoft.ServiceFabric.Actors.Runtime
             this.actor = actor;
             this.stateProvider = actorStateProvider;
             this.stateChangeTracker = new Dictionary<string, StateMetadata>();
+            this.clock = new SystemClock();
```
We should inject IClock instead of creating new instances in different classes. Although we can use Inspector/Reflection to manipulate the field, expressing the dependency explicitly, as a constructor parameter is better.

olegsych commented on src/Microsoft.ServiceFabric.Actors/Runtime/ActorManager.cs:
```diff
@@ -34,26 +35,41 @@ internal sealed class ActorManager : IActorManager
+        private readonly IClock clock;
+        private readonly PerformanceCounterProviderV2 performanceCounterProvider;
+
+        private static Func<ActorService, IClock, PerformanceCounterProviderV2, IDiagnosticEvents> createDiagnosticEvents = (actorService, clock, performanceCounterProvider) =>
```
I think this logic better fits in the ActorService class because it also creates the ActorStateManager, which also needs IDiagnosticEvents. So, having it here introduces unnecessary coupling. For example, ActorManager creates and disposes the IDiagnosticEvents, but ActorStateManager uses it and that dependency is not visible here. This class could decide to dispose and recreate it breaking the ActorStateManager. However, if the ActorService creates an instance and gives it to both ActorManager and ActorStateManager, the intent would be clear.

ActorService is also a better fit for the role of the *composition root*, so I would expect dependencies like IDiagnosticEvents, IClock, etc to bubble up to it.

ilijagrbic replied:
Thanks for explaining this. I was biased when thinking about this since the original location for old `DiagnosticManager` was in the `ActorManager`. Anyways, I think this should be tackled (together with the `IClock` issue) and I will do it in a follow up PR.

olegsych commented on src/Microsoft.ServiceFabric.Actors/Diagnostics/IDiagnosticEvents.cs:
```diff
@@ -0,0 +1,43 @@
+
+namespace Microsoft.ServiceFabric.Actors.Diagnostics
+{
+    internal interface IDiagnosticEvents
```
After commenting on the name of the ActorBase.DiagnosticsContext property, it occurred to me that the term Events is also redundant. So the interface name could be simplified to *IDiagnostics* without losing descriptiveness.

ilijagrbic replied:
I like how this is an example of "less is more" when it comes to naming things. We converged to a one-word name after thinking about way more complicated names in the beginning.

olegsych commented on test/Microsoft.ServiceFabric.Actors.Tests/Runtime/ActorManagerTests.cs:
```diff
@@ -0,0 +1,345 @@
+
+namespace Microsoft.ServiceFabric.Actors.Runtime
+{
+    public class ActorManagerTests
```
Test class names should be singular - ActorManagerTest, otherwise the test names don't make sense in English.

ilijagrbic replied:
I completely missed this since it was a preexisting class. I will rename it.

olegsych commented on test/Microsoft.ServiceFabric.Actors.Tests/ActorServiceRemotingDispatcherTest.cs:
```diff
@@ -0,0 +1,67 @@
+
+            public DiagnosticEvents()
+            {
+                var method = sut.GetType().GetMethod("HandleActorMethodDispatchAsync",
```
Ugh. Can't we use Inspector to make this more compact?

Something along the lines of https://github.com/olegsych/inspector/blob/de276bbfb1357bf9415bd35f07daeb7b138c88a5/examples/ObjectAccessExample.cs#L104?

ilijagrbic replied:
Initially I had some issue with implementing the Inspector approach here, so I settled to just use reflection. I managed to get it to work now, so I will implement this suggestion. :)

olegsych commented on test/Microsoft.ServiceFabric.Actors.Tests/AssemblyInfo.cs:
```diff
@@ -3,10 +3,16 @@
 [assembly: CollectionBehavior(CollectionBehavior.CollectionPerAssembly)]
 [assembly: CaptureConsole]
+
+// Used to test mocks
```
Although it's not immediately obvious that DynamicProxyGenAssembly is used for mocking, it's pretty close; plus we have a doc comment for the constant explaining this. I would remove the line comment.

olegsych reviewed (approved):
I've completed another review pass @ilijagrbic. Aside from naming refinements, my main suggestion is to use ActorService as the composition root and create the IDiagnosticEvents objects there instead of the ActorManager and express dependencies explicitly with constructor parameters. The PR already feels too large, so feel free to merge it and iterate.

ilijagrbic committed:
Address PR comments

ilijagrbic committed:
Address PR comments

ilijagrbic committed:
Delete ActorMethodDiagnosticData constructor

ilijagrbic committed:
Revert accidental change

ilijagrbic committed:
Replace reflection with Inspector for ActorServiceTest

olegsych reviewed (commented):
:shipit:
